Mage.Sets/src/mage/cards/f/FirestormPhoenix.java

```diff
@@ -0,0 +1,203 @@
+        this.addAbility(new SimpleStaticAbility(Zone.ALL, new FirestormPhoenixEffect()));
+    }
+
+    private FirestormPhoenix(final FirestormPhoenix card) {
+        super(card);
+    }
+
+    @Override
+    public FirestormPhoenix copy() {
+        return new FirestormPhoenix(this);
+    }
+}
+
+class FirestormPhoenixEffect extends ReplacementEffectImpl {
+	
+	FirestormPhoenixEffect() {
+		super(Duration.Custom, Outcome.ReturnToHand);
+		staticText = "If {this} would die, return {this} to its owner's hand instead. Until that player’s next turn, that player plays with that card revealed in their hand and can’t play it.";
+	}
+	
+	private FirestormPhoenixEffect(final FirestormPhoenixEffect effect) {
+		super(effect);
+	}
+	
+	@Override
+	public FirestormPhoenixEffect copy() {
+		return new FirestormPhoenixEffect(this);
+	}
+	
+	// First attempt to write FirestormPhoenix effect
+	@Override
+	public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+	    Player controller = game.getPlayer(source.getControllerId());
+	    if (controller != null) {
+	        Permanent permanent = ((ZoneChangeEvent) event).getTarget();
+	        Cards revealedCards = new CardsImpl();
+	        game.getCard(permanent.getId());
+	        revealedCards.add(game.getCard(((ZoneChangeEvent) event).getTargetId()));
+	        if (permanent != null) {
+	        	
+	        	// Create ContinuousRuleModifyingEffect that restricts the casting of this card, until your next turn.
+	        	ContinuousEffect effectRestrict = new FirestormPhoenixRestrictEffect(permanent, game);
+	            game.addEffect(effectRestrict, source);
+	            
+	            // Create ContinuousEffectImpl that reveals FirestormPhoenix from hand continually, 
+	            // until it moves zone (is not in your hand) or until your next turn, whichever comes first. Need to add custom inActive.
+	            // Use game.informPlayers to inform players when the card changes zones.
+	            ContinuousEffect effectReveal = new FirestormPhoenixRevealEffect(controller, revealedCards, permanent.getIdName());
+	            game.addEffect(effectReveal, source);
+	        	
+	            // Move this card (FirestormPhoenix) to hand.
+	        	controller.moveCards(permanent, Zone.HAND, source, game);
+	        
+	            return true;
+	        }
+	    }
+	    return false;
+	}
+	
+    @Override
+    public boolean checksEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.ZONE_CHANGE;
+    }
+	
+    @Override
+    public boolean applies(GameEvent event, Ability source, Game game) {
+    	return source.getSourceId().equals(event.getTargetId()) 
+        && ((ZoneChangeEvent) event).isDiesEvent();
+    }
+}
+
+class FirestormPhoenixRevealEffect extends ContinuousEffectImpl {
+	
+	private Player controller;
+	private Cards revealedCards;
```

Comment:
this works I guess, but could be better to use the existing target functionality rather than passing in custom cards param- when creating the effect, use `.setTargetPointer(new FixedTarget(xxx, game))` and then `getTargetPointer().getFirst(game.source)`, this automatically checks zcc

Reply:
That's much better, thank you!